### PR TITLE
Scan POMs and JARs for license information

### PR DESCRIPTION
**Pull request by dwalluck**

### CODE REVIEW

**Review by vibe13 — CHANGES_REQUESTED** [review]
Hi @dwalluck thanks a lot for this massive work! I have put some comments around, mostly because I missed some explanation to understand some method and it took some time to figure it all out. I have run an analysis on a big repo (jboss-eap-7.4.15.GA-CR1-maven-repository), and checked a few licenses. They mostly make sense but in some cases some GAVs are reported as missing license information, but looking at their POMs, the license seems ok to be picked. I will give you details privately so you can debug yourself.
Lastly, I still fail to see logs in the console when executing the jar :-/.

Thanks!

File: core/src/main/java/org/jboss/pnc/build/finder/core/DistributionAnalyzer.java [hunk 1]
```diff
@@ -557,6 +599,111 @@ private void listChildren(FileObject fo) throws IOException {
...
+                }
+
+                if (name != null) {
+                    name = SPACE_PATTERN.matcher(name).replaceAll(" ");
+                }
+
+                String licenseId;
+                Optional<String> optLicenseId = LicenseUtils.findMatchingLicense(name, url);
+
+                if (optLicenseId.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (dwalluck) [reply core/src/main/java/org/jboss/pnc/build/finder/core/DistributionAnalyzer.java:631]: This increases the number of `if` checks, but I just moved Build Finder to Java 11 and discovered `Optional::or` from Java 9+, so I can use this I think:

```java
Optional<String> optLicenseId = LicenseUtils.findMatchingLicense(name, url)
                        .or(() -> LicenseUtils.findLicenseMapping(mapping, url))
                        .or(() -> LicenseUtils.findLicenseMapping(mapping, name));
```

File: core/src/main/java/org/jboss/pnc/build/finder/core/LicenseUtils.java [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    private static final String[] EXTENSIONS_TO_REMOVE = { ".html", ".php", ".txt" };
+
+    private static final Pattern NAME_VERSION_PATTERN = Pattern
+            .compile("(?<name>[A-Z-a-z])[Vv]?(?<major>[1-9]+)(\\.(?<minor>[0-9]+))?");
+
+    private LicenseUtils() {
+        throw new IllegalArgumentException("This is a utility class and cannot be instantiated");
+    }
+
+    private static Map<String, SpdxListedLicense> getLicenseIdMap() {
```
[the remark this answers is not in the document]
Comment (dwalluck) [reply core/src/main/java/org/jboss/pnc/build/finder/core/LicenseUtils.java:88]: What's annoying is that the official SPDX Java API stores a map internally, but only returns a list (`map.values()`) in the public API.

File: core/src/main/java/org/jboss/pnc/build/finder/core/DistributionAnalyzer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vibe13) [reply core/src/main/java/org/jboss/pnc/build/finder/core/DistributionAnalyzer.java:630]: nice!

File: core/src/main/java/org/jboss/pnc/build/finder/core/LicenseUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vibe13) [reply core/src/main/java/org/jboss/pnc/build/finder/core/LicenseUtils.java]: oh yes that would have been handy. but one initialization is better now, thanks!

**Review by vibe13 — APPROVED** [review]
Massive work David, thanks a lot!!

**Pull request merged**